File: Day3/Day3.java
import java.io.File;
import java.io.FileNotFoundException;
import java.util.ArrayList;
import java.util.Scanner;

class Day3 {

    public static int getPriority(char character) {
        // Luckily ords in java are similar to python
        int originalOrd = (int) character;
        if (Character.isLowerCase(character)) {
            return originalOrd - 96;
        }
        return originalOrd - 38; // 64 - 26
    }

    public static char getItem(String rucksack) {
        // Retrieving the correct item type
        String first = rucksack.substring(0, (int) rucksack.length() / 2);
        String second = rucksack.substring((int) rucksack.length() / 2, rucksack.length());
        for (int i = 0; i < first.length(); i++) {
            for (int j = 0; j < second.length(); j++) {
                if (((Character) first.charAt(i)).equals((Character) second.charAt(j))) {
                    return first.charAt(i);
                }
            }
        }
        return 'a'; // compiler yelling at me for not always specifying a return
    }

    public static char getBadge(ArrayList<String> rucksack) {
        // Retrieving the badge from a group of 3 little elves
        for (int i = 0; i < rucksack.get(0).length(); i++) {
            for (int j = 0; j < rucksack.get(1).length(); j++) {
                for (int k = 0; k < rucksack.get(2).length(); k++) {
                    if (((Character) rucksack.get(0).charAt(i)).equals((Character) rucksack.get(1).charAt(j)) && ((Character) rucksack.get(0).charAt(i)).equals((Character) rucksack.get(2).charAt(k))) {
                        return rucksack.get(0).charAt(i);
                    }
                }
            }
        }
        return 'a';
    }

    public static int problem1(String filename) throws FileNotFoundException
    {
        File f = new File(filename);
        Scanner s = new Scanner(f);
        try
        {
            ArrayList<String> lines = new ArrayList<String>();
            while (s.hasNextLine())
            {
                lines.add(s.nextLine().strip());
            }

            // Yay, time to do this again
            // I don't think java has a way to define methods inside methods so they'll be at the top
            int total = 0;
            for (String line: lines) {
                total += getPriority(getItem(line));
            }
            
            return total;
        }
        finally
        {
            s.close();
        }
    }

    public static int problem2(String filename) throws FileNotFoundException
    {
        File f = new File(filename);
        Scanner s = new Scanner(f);
        try
        {
            ArrayList<String> lines = new ArrayList<String>();
            while (s.hasNextLine())
            {
                lines.add(s.nextLine().strip());
            }

            // Did what I did in python...
            int total = 0;
            ArrayList<String> temp = new ArrayList<String>();

            for (String line: lines) {
                temp.add(line);
                if (temp.size() == 3) {
                    total += getPriority(getBadge(temp));
                    temp = new ArrayList<String>();
                }
            }
            
            return total;
        }
        finally
        {
            s.close();
        }
    }
    
    public static void main(String[] args) {
        try {
            System.out.println(problem1("data.txt"));
            System.out.println(problem2("data.txt"));
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}

File: Day3/Day3.py
def problem1(file: str) -> int:
    """
    Problem 1 Solution"""
    lines = []
    with open(file, "r") as f:
        for line in f:
            lines.append(line.strip())

    # First let's find the only items that are in both compartments
    def find_item(compartment: str) -> str:
        """
        Find the item that appears in the compartment thing twice

        We can just use some iteration to find it
        """
        over, under = (
            compartment[: round(len(compartment) / 2)],
            compartment[round(len(compartment) / 2) :],
        )
        for char in over:
            for second_char in under:
                if char == second_char:
                    return char

    def get_priority(item: str) -> int:
        """
        Return the priority number

        The best way to go about this is probably ordinals
        """
        return ord(item) - 96 if item.islower() else ord(item) - 38  # 64 - 26

    total = 0
    for line in lines:
        total += get_priority(find_item(line))

    f.close()
    return total


def problem2(file: str) -> int:
    """
    Problem2 Solution
    """
    lines = []
    with open(file, "r") as f:
        for line in f:
            lines.append(line.strip())

    # Here we just have to modify our find function by a bit, let's rewrite it
    def find_badge(lines: list) -> str:
        """
        Find the badge of a group of 3 elves
        """
        for char in lines[0]:
            for second_char in lines[1]:
                for third_char in lines[2]:
                    if char == second_char and char == third_char:
                        return char

    # Luckily we can just use the same get_priority function
    def get_priority(item: str) -> int:
        """
        Return the priority number
        """
        return ord(item) - 96 if item.islower() else ord(item) - 38

    # Now we just have to provide all the correct items to the functions
    total = 0
    temp = []

    for line in lines:
        temp.append(line)
        if len(temp) == 3:
            total += get_priority(find_badge(temp))
            temp = []

    f.close()
    return total


print(problem1("data.txt"))
print(problem2("data.txt"))
